packages/plugin-hardhat/src/validate-upgrade.ts

```diff
@@ -0,0 +1,57 @@
+import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { ContractFactory } from 'ethers';
+
+import { ContractAddressOrInstance, getContractAddress } from './utils';
+import {
+  getBeaconAddress,
+  isBeaconProxy,
+  isTransparentOrUUPSProxy,
+  isBeacon,
+  ValidateUpgradeUnsupportedError,
+  assertUpgradeSafe,
+  assertStorageUpgradeSafe,
+  ValidationOptions,
+} from '@openzeppelin/upgrades-core';
+import { validateBeaconImpl, validateProxyImpl } from './utils/validate-impl';
+import { getDeployData } from './utils/deploy-impl';
+
+export interface ValidateUpgradeFunction {
+  (origImplFactory: ContractFactory, newImplFactory: ContractFactory, opts?: ValidationOptions): Promise<void>;
+  (
+    proxyOrBeaconAddress: ContractAddressOrInstance,
+    newImplFactory: ContractFactory,
+    opts?: ValidationOptions,
+  ): Promise<void>;
+}
+
+export function makeValidateUpgrade(hre: HardhatRuntimeEnvironment): ValidateUpgradeFunction {
+  return async function validateUpgrade(
+    addressOrImplFactory: ContractAddressOrInstance | ContractFactory,
+    newImplFactory: ContractFactory,
+    opts: ValidationOptions = {},
+  ) {
+    if (addressOrImplFactory instanceof ContractFactory) {
+      const newDeployData = await getDeployData(hre, newImplFactory, opts);
+      assertUpgradeSafe(newDeployData.validations, newDeployData.version, newDeployData.fullOpts);
+
+      const origDeployData = await getDeployData(hre, addressOrImplFactory, opts);
+      if (opts.unsafeSkipStorageCheck !== true) {
+        assertStorageUpgradeSafe(origDeployData.layout, newDeployData.layout, newDeployData.fullOpts);
+      }
+    } else {
+      const proxyOrBeaconAddress = getContractAddress(addressOrImplFactory);
+      const { provider } = hre.network;
+      const deployData = await getDeployData(hre, newImplFactory, opts);
+      if (await isTransparentOrUUPSProxy(provider, proxyOrBeaconAddress)) {
+        await validateProxyImpl(deployData, opts, proxyOrBeaconAddress);
+      } else if (await isBeaconProxy(provider, proxyOrBeaconAddress)) {
+        const beaconAddress = await getBeaconAddress(provider, proxyOrBeaconAddress);
+        await validateBeaconImpl(deployData, opts, beaconAddress);
+      } else if (await isBeacon(provider, proxyOrBeaconAddress)) {
+        await validateBeaconImpl(deployData, opts, proxyOrBeaconAddress);
+      } else {
+        throw new ValidateUpgradeUnsupportedError(proxyOrBeaconAddress);
```

Comment:
It really sounds like we should be using the same Error instance here and in `prepareUpgrade` (and potentially in other places?).

I get that it's currently named `PrepareUpgradeUnsupportedError`, but we can rename it, add `export const PrepareUpgradeUnsupportedError = __newName__` for backwards compat.

Reply:
Instead of throwing an error, we could just support this condition where we validate an upgrade where the original implementation address was passed in as a parameter (as long as the impl is in the manifest).  Changed it to do this.